File: Numeros Pseudo-Aleatorios/Linear Congruential/LinearCongruential.go
/*
 PRNG .......: Linear Congruential
 Ano ........: 1965
 Autor ......: W. E. Thomson
*/

package main

import (
    "fmt"
    "time"
)

func linearCongruential(seed *int64) float64 {
    const M = 2147483647
    *seed = (1664525 * *seed + 1013904223) % M
    return float64(*seed) / float64(M)
}

func main() {
    var Xn1 int64 = 1234
    Xn2 := time.Now().UnixNano() / int64(time.Millisecond)

    fmt.Println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    fmt.Println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    fmt.Println("PADRAO: Linear Congruential")
    fmt.Println("==========================================")
    fmt.Println()
    fmt.Println("    Semente: 1234        Semente: TIME")
    fmt.Println("    -------------        -------------")

    for i := 1; i <= 10; i++ {
        fmt.Printf("%17.11f        %13.11f\n", linearCongruential(&Xn1), linearCongruential(&Xn2))
    }
}

File: Numeros Pseudo-Aleatorios/Linear Congruential/LinearCongruential.pas
{ 
 PRNG .......: Linear Congruential
 Ano ........: 1965
 Autor ......: W. E. Thomson
}

program Linear_Congruential;

uses
  SysUtils, DateUtils;
  
function LinearCongruential(var seed: Int64): Double;
const
  M = 2147483647;
begin
  seed := (1664525 * seed + 1013904223) mod M;
  LinearCongruential := seed / M;
end;

var
  Xn1, Xn2: Int64;
  i: Integer;

begin

  Xn1 := 1234; 
  Xn2 := LongWord(DateTimeToUnix(Now)); 

  writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  writeln('PADRAO: Linear Congruential');
  writeln('==========================================');
  writeln;
  writeln('    Semente: 1234        Semente: TIME');
  writeln('    -------------        -------------');

  for i := 1 to 10 do
    writeln(LinearCongruential(Xn1):17:11, '        ', LinearCongruential(Xn2):13:11);
    
end.
